### PR TITLE
sl-141: Add feedback section to site

### PR DESCRIPTION
**Pull request by dimapopovych**
image
image

### CODE REVIEW

File: backend/src/libs/packages/open-ai/open-ai.package.ts [hunk 1]
```diff
@@ -37,6 +37,14 @@ class OpenAI {
     return this.parseCompletionResponse(text);
   }
 
+  public async createImage(prompt: string): Promise<string> {
+    const response = await this.openAIApi.createImage({
+      prompt,
+    });
+
+    return response.data.data[0]?.url ?? '';
```
Comment (what1s1ove) [inline backend/src/libs/packages/open-ai/open-ai.package.ts]:
```suggestion
    const { data } = await this.openAIApi.createImage({
      prompt,
    });
    
    const [image] = data.data;

    return image?.url ?? '';
```
but you need to get an buffer i guess to save it to the s3

File: backend/src/packages/sections/section.service.ts [hunk 2]
```diff
@@ -1,3 +1,5 @@
+import { initAsyncItemsQueue } from 'shared/build/index.js';
```
Comment (what1s1ove) [inline backend/src/packages/sections/section.service.ts]: import

File: frontend/src/libs/components/carousel/carousel.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import React, { useCallback, useState } from 'react';
+
+import { getValidClassNames } from '~/libs/helpers/helpers.js';
+
+import styles from './styles.module.scss';
+
+const Carousel: React.FC<React.PropsWithChildren> = ({
+  children,
+}: React.PropsWithChildren) => {
+  const [currentSlide, setCurrentSlide] = useState(0);
```
Comment (what1s1ove) [inline frontend/src/libs/components/carousel/carousel.tsx]: type state

File: frontend/src/libs/components/carousel/carousel.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import LibraryCarousel, { type ControlProps } from 'nuka-carousel';
+import React, { useCallback, useState } from 'react';
```
Comment (what1s1ove) [inline frontend/src/libs/components/carousel/carousel.tsx]: hooks import

File: frontend/src/pages/site/components/feedback/feedback.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    name: string;
+    profession: string;
+    feedback: string;
+  }[][] = [];
+
+  for (let index = 0; index < cards.length; index += 2) {
+    slidesContent.push(cards.slice(index, index + 2));
+  }
+
+  const handleRenderSlides = (): JSX.Element[] => {
```
Comment (what1s1ove) [inline frontend/src/pages/site/components/feedback/feedback.tsx]: move it to the jsx

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/libs/packages/open-ai/open-ai.package.ts [hunk 6]
```diff
@@ -37,6 +38,21 @@ class OpenAI {
...
 
+  public async createImage(prompt: string): Promise<string> {
+    const { data } = await this.openAIApi.createImage({
+      prompt,
+      response_format: 'b64_json',
+    });
+
+    const [image] = data.data;
+
+    const buffer = Buffer.from(image?.b64_json ?? '', 'base64');
```
Comment (dimapopovych) [inline backend/src/libs/packages/open-ai/open-ai.package.ts]: Maybe should move it to file package

File: frontend/src/pages/site/components/feedback/feedback.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        <div className={styles['feedback-carousel']}>
+          <Carousel>
+            {cards.map((card, index, array) => {
+              if (index % 2 === 0) {
+                const nextCard = array[index + 1] as {
+                  photo: string;
+                  name: string;
+                  profession: string;
+                  feedback: string;
+                };
```
Comment (dimapopovych) [inline frontend/src/pages/site/components/feedback/feedback.tsx]: FeedbackCardContent type? But I don't where I should declare it

File: backend/src/libs/packages/open-ai/open-ai.package.ts (hunk 6 above)
Comment (what1s1ove) [reply backend/src/libs/packages/open-ai/open-ai.package.ts:57]: agreed

File: backend/src/libs/packages/open-ai/libs/maps/image-size-to-resolution.map.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import { CreateImageRequestSizeEnum } from 'openai';
+
+import { ImageSize } from '~/libs/packages/open-ai/libs/enums/enums.js';
+import { type ValueOf } from '~/libs/types/types.js';
+
+const ImageSizeToResolutionMap: Record<
```
Comment (what1s1ove) [inline backend/src/libs/packages/open-ai/libs/maps/image-size-to-resolution.map.ts]:
```suggestion
const imageSizeToResolutionMap: Record<
```

File: backend/src/libs/packages/open-ai/open-ai.package.ts (hunk 6 above)
Comment (what1s1ove) [reply backend/src/libs/packages/open-ai/open-ai.package.ts]: i think we will use it only in this place. at lease for now. let's

File: backend/src/packages/sections/section.service.ts (hunk 2 above)
Comment (what1s1ove) [reply backend/src/packages/sections/section.service.ts]: isn't resolved. should be from `libs/helpers/helpers`

File: backend/src/libs/packages/open-ai/open-ai.package.ts [hunk 9]
```diff
@@ -37,6 +42,25 @@ class OpenAI {
...
+      prompt,
+      size: ImageSizeToResolutionMap[size],
+      response_format: 'b64_json',
+    });
+
+    const [image] = data.data;
+
+    const buffer = Buffer.from(image?.b64_json ?? '', 'base64');
+
+    const { url } = await file.upload({ file: buffer });
```
Comment (what1s1ove) [inline backend/src/libs/packages/open-ai/open-ai.package.ts]:
```suggestion
    const { url } = await this.file.upload({ file: buffer });
```

File: frontend/src/pages/site/components/feedback/feedback.tsx [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+type Properties = {
+  content: SiteFeedbackContent;
+};
+
+import styles from './styles.module.scss';
+
+const Feedback: React.FC<Properties> = ({
+  content: { title, cards },
+}: Properties) => {
+  const handleRenderTitle = (): JSX.Element => {
```
Comment (what1s1ove) [inline frontend/src/pages/site/components/feedback/feedback.tsx]: avoid render functions. move everything to jsx + variables

File: frontend/src/pages/site/components/feedback/feedback.tsx [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+  return (
+    <div className={styles['feedback']}>
+      <div className={styles['feedback-container']}>
+        {handleRenderTitle()}
+        <div className={styles['feedback-carousel']}>
+          <Carousel cellSpacing={28}>
+            {cards.map((card, index, array) => {
+              if (index % 2 === 0) {
+                const nextCard = array[index + 1] as {
```
Comment (what1s1ove) [inline frontend/src/pages/site/components/feedback/feedback.tsx]: too complicated

consider using this library instead - https://www.npmjs.com/package/react-multi-carousel

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/pages/site/components/feedback/feedback.tsx (hunk 11 above)
Comment (dimapopovych) [reply frontend/src/pages/site/components/feedback/feedback.tsx]: I think solved. Now looks like this
```typescript
          <Carousel slidesToShow={2} slidesToScroll={2} cellSpacing={28}>
            {cards.map((card) => (
              <FeedbackCard card={card} key={card.photo} />
            ))}
          </Carousel>
```

File: backend/src/libs/packages/open-ai/open-ai.package.ts [hunk 12]
```diff
@@ -37,6 +51,25 @@ class OpenAI {
...
+      prompt,
+      size: imageSizeToResolutionMap[size],
+      response_format: 'b64_json',
+    });
+
+    const [image] = data.data;
+
+    const buffer = Buffer.from(image?.b64_json ?? '', 'base64');
+
+    const { url } = await this.file.upload({ file: buffer });
```
Comment (what1s1ove) [inline backend/src/libs/packages/open-ai/open-ai.package.ts]: actually ai service shouldn't know anything about file saving. it just need to return text/image

move file service to place where you really need it. always try to decompose the logic

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/packages/sections/section.service.ts [hunk 13]
```diff
@@ -106,14 +128,48 @@ class SectionService
...
+
+  private async createFeedbackContent(
+    prompt: string,
+  ): Promise<SiteFeedbackContent> {
+    const CARDS_COUNT = 8;
+
+    const cardsContent = await Promise.all(
+      Array.from({ length: CARDS_COUNT }).map(() =>
+        openAI.createCompletion(prompt),
+      ),
```
Comment (what1s1ove) [inline backend/src/packages/sections/section.service.ts]:
```suggestion
      Array.from(new Array(SectionService .CARDS_COUNT), () => openAI.createCompletion(prompt))
```
add CARDS_COUNT to static field to `SectionService` (give a better name for it)

2 argument of `Array.from` is a regular `arr.map`. prefer it

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/pages/site/components/feedback/feedback.tsx (hunk 11 above)
Comment (what1s1ove) [reply frontend/src/pages/site/components/feedback/feedback.tsx]: okay, got it

File: backend/src/packages/sections/section.service.ts (hunk 13 above)
Comment (dimapopovych) [reply backend/src/packages/sections/section.service.ts]:
```suggestion
      Array.from({ length: SectionService.feedbackCardsQuantity }, () => openAI.createCompletion(prompt))
```
Due to linter

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**